scvi/inference/total_inference.py

```diff
@@ -618,13 +1052,97 @@ def loss(self, tensors):
             + self.pro_recons_weight * reconst_loss_protein
             + self.kl_weight * kl_div_z
             + kl_div_l_gene
-            + self.back_warmup_weight * kl_div_back_pro
+            + self.kl_weight * kl_div_back_pro
         )
+
         return loss
 
-    def on_epoch_begin(self):
-        super().on_epoch_begin()
-        if self.n_epochs_back_kl_warmup is not None:
-            self.back_warmup_weight = min(1, self.epoch / self.n_epochs_back_kl_warmup)
+    def loss_discriminator(
+        self, z, batch_index, predict_true_class=True, return_details=True
+    ):
+
+        n_classes = self.gene_dataset.n_batches
+        cls_logits = torch.nn.LogSoftmax(dim=1)(self.discriminator(z))
+
+        if predict_true_class:
+            cls_target = one_hot(batch_index, n_classes)
+        else:
+            one_hot_batch = one_hot(batch_index, n_classes)
+            cls_target = torch.zeros_like(one_hot_batch)
+            # place zeroes where true label is
+            cls_target.masked_scatter_(
+                ~one_hot_batch.bool(), torch.ones_like(one_hot_batch) / (n_classes - 1)
+            )
+
+        l_soft = cls_logits * cls_target
+        loss = -l_soft.sum(dim=1).mean()
+
+        return loss
+
+    def _get_z(self, tensors):
+        (
+            sample_batch_X,
+            local_l_mean,
+            local_l_var,
+            batch_index,
+            label,
```

Comment:
is this function necessary ?

Reply:
that code will need to be somewhere, be it in the `on_training_loop` or as a "private" function here.

Reply:
ok